test(gnovm): "*_filetest.gno" instructions

tbruyelle opened the pull request:
Add tests for Output, Error and Realm instructions

The change introduces a new test dependency:  [testscripts][0], which allows to create tests based on [txtar][1] files. This is very handy for testing `gno test`, because:
- you can easily creates a set of files that can be used for a command (here the `gno test` command)
- you can assert stderr and stdout expectations in a very readable and maintainable way
- you can compare files using the `cmp` command, which is very useful to assert of the `-update-golden-tests` behavior.

The tests can be run with the following command:
```sh
$ go test ./gnovm/tests/ -v -run TestRunFileTest
```

These tests were added to help work on #1015, which will potentially alter the behavior of "*_filetest.gno" instructions updates.

<!-- please provide a detailed description of the changes made in this pull request. -->

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

[0]: https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript
[1]: https://pkg.go.dev/golang.org/x/tools/txtar

tbruyelle commented on gnovm/tests/testdata/error_incorrect.txt:
```diff
@@ -0,0 +1,17 @@
+# Test Error instruction incorrect
+
+! exec $GNO test -verbose -root-dir=$ROOTDIR .
```
The leading `!` indicates the command is expected to fail (exit code != 0)

tbruyelle commented on gnovm/tests/testdata/error_correct.txt:
```diff
@@ -0,0 +1,18 @@
+}
+
+// Error:
+// oups
```
Each txtar file (txt is also an accepted extension) is executed in his own directory, so here the only file present for that test is `x_filetest.gno`.

moul replied:
It looks great.

What do you think about moving this to `gnovm/cmd/gno/testdata`, since it’s testing the CLI?

tbruyelle replied:
Good point. Ideally, I think those test should be added to `gnovm/cmd/gno/test_test.go`, and so this whole file should be rewritten using `testscripts`. Indeed I find the current format hard to read and maintain. The only good part is it fits in a single file, whereas with typescript there would be multiple small txtar files.

tbruyelle replied:
> Good point. Ideally, I think those tests should be added to `gnovm/cmd/gno/test_test.go`, and so this whole file should be rewritten using `testscripts`. Indeed I find the current format hard to read and maintain. The only good part is it fits in a single file, whereas with typescript there would be multiple small txtar files.

tbruyelle commented on gnovm/tests/testdata/output_correct.txt:
```diff
@@ -0,0 +1,20 @@
+
+exec $GNO test -verbose -root-dir=$ROOTDIR .
+
+! stdout .+
```
Means we expect nothing from stdout.

moul replied:
Can your useful GitHub comments be real file comments?

tbruyelle replied:
sure ! 
f657673

moul commented on gnovm/tests/file_test.go:
```diff
@@ -100,3 +104,29 @@ func runFileTest(t *testing.T, path string, opts ...RunFileTestOption) {
+	rootDir := path.Dir(string(goModPath))
+	// Build a fresh gno binary in a temp directory
+	gnoBin := path.Join(t.TempDir(), "gno")
+	err = exec.Command("go", "build", "-o", gnoBin, path.Join(rootDir, "gnovm", "cmd", "gno")).Run()
```
```suggestion
	err = exec.Command("go", "build", "-o", gnoBin, filepath.Join(rootDir, "gnovm", "cmd", "gno")).Run()
```

tbruyelle replied:
done here 
f657673

harry-hov commented on gnovm/tests/file_test.go:
```diff
@@ -100,3 +104,29 @@ func runFileTest(t *testing.T, path string, opts ...RunFileTestOption) {
+	// Get root location of github.com/gnolang/gno
+	goModPath, err := exec.Command("go", "env", "GOMOD").CombinedOutput()
+	require.NoError(t, err)
+	rootDir := path.Dir(string(goModPath))
```
maybe?
```suggestion
	rootDir := filepath.Dir(string(goModPath))
```

tbruyelle replied:
You're probably right, I assume that `go env GOMOD` returns a path in the expected OS format, so `filepath` is better! [c3d79a3](https://github.com/gnolang/gno/pull/1023/commits/c3d79a3c3909e67f173388a2b723e63eef86b4f0)

harry-hov commented on gnovm/tests/file_test.go:
```diff
@@ -100,3 +104,29 @@ func runFileTest(t *testing.T, path string, opts ...RunFileTestOption) {
+	require.NoError(t, err)
+	rootDir := path.Dir(string(goModPath))
+	// Build a fresh gno binary in a temp directory
+	gnoBin := path.Join(t.TempDir(), "gno")
```
maybe?

```suggestion
	gnoBin := filepath.Join(t.TempDir(), "gno")
```

tbruyelle replied:
right too 
6a2ed63

harry-hov reviewed (approved):
LGTM 💯 

Really welcoming `testscript` package :shipit:

zivkovicmilos reviewed (approved):
I think this is really cool

Thank you for this addition 🙏

moul reviewed (changes requested):
Can you address this before: https://github.com/gnolang/gno/pull/1023#discussion_r1285598833, please?

tbruyelle commented:
> Can you address this before: [#1023 (comment)](https://github.com/gnolang/gno/pull/1023#discussion_r1285598833), please?

@moul the last commits address that : 

- moved `gnovm/tests.RunFileTest` in `gnovm/cmd/gno.TestTest`
- migrate all cases in `TestTest` so they use testscript.
Some of them were using files in `examples`, I replaced those cases by `testscript` file so we don't take the risk of breaking the tests when changing files in `examples`.
- clean gno files in `gnovm/tests/integ` that aren't used any more (due to previous point)

Regarding `testscript` setup, I also provided a couple of improvements:
- replace `exec $GNO` with a simple `gno`, which is configured as a `testscript` command/keyword that targets the gno binary compiled during setup.
- moved the setup in a method in `main_test.go`, so when we'll migrate the other `gno` commands it will be easier.
- replace `-root-dir` flag by `$GNOROOT`, thanks to #1014 and a rebase on master.
